torch/jit/__init__.py

```diff
@@ -2079,6 +2079,80 @@ def _get_script_class(name):
                            "Did you forget to import it?".format(name))
     return _script_classes[name]
 
+# qualified_name => list[(overload decl, overload default args)]
+_overloaded_fns = {}
+
+# qualified name => list[compiled fns]
+_compiled_overloaded_fns = {}
+
+def overload(func):
+    qual_name = _qualified_name(func)
+    global _overloaded_fns
+    li = _overloaded_fns.get(qual_name)
+    if li is None:
+        li = []
+        _overloaded_fns[qual_name] = li
+    signature = torch.jit.annotations.get_signature(func)
+    if signature is None:
+        raise RuntimeError("Must explicitly add type annotations to overloaded functions: {obj}").format(func)
+    li.append((torch.jit.get_jit_def(func).decl(), get_default_args(func)))
+    return func
+
+def compile_function_with_overload(qual_name, impl_fn, overload_decl, overload_defaults):
+    _frames_up = 0
+    impl_ast = torch.jit.get_jit_def(impl_fn)
+    # TODO: figure out correct closure_rcb, stack_rcb, although this seems to work
+    # and refactor with above usage
+    closure_rcb = _jit_internal.createResolutionCallbackFromClosure(impl_fn)
+    stack_rcb = _jit_internal.createResolutionCallback(_frames_up + 1)
+
+    def _rcb(name):
+        # since type comments aren't captured in the function's closures,
+        # we still need to try to the rcb based on stack frames if the
+        # closure rcb fails
+        result = closure_rcb(name)
+        if result:
+            return result
+        return stack_rcb(name)
+    fn = torch._C._jit_script_compile_overload(qual_name, overload_decl, impl_ast, _rcb, overload_defaults)
+    return fn
+
+def get_overloads(obj):
+    # check for cached compiled fns
+    qual_name = _qualified_name(obj)
+    global _compiled_overloaded_fns
+    compiled_overloads = _compiled_overloaded_fns.get(qual_name, None)
+    if compiled_overloads is not None:
+        return compiled_overloads
+
+    # check for not yet compiled overloads
+    global _overloaded_fns
+    overloads = _overloaded_fns.get(qual_name, None)
+    if overloads is None:
+        return None
+    compiled_fns = []
+    # TODO: use default args from the implementation, not from the overload
+    # This is more complicated because you can have a default arg with a type
+    # incompatible with a type of parameter in an overload, and other validation.
+    # This is still an internal api so for now use defaults from overload
+    for overload_decl, overload_defaults in overloads:
+        compiled_fn = compile_function_with_overload(qual_name, obj, overload_decl, overload_defaults)
+        compiled_fns.append(compiled_fn)
+
+    # cache compilation, remove information stored to do compilation
+    _compiled_overloaded_fns[qual_name] = compiled_fns
```

Comment:
This looks like a potential memory leak, if the functions this compiles go out of scope, their compiled versions will stick around here. @Chillee found a similar issue recently, I don't remember if he found a fix or not

Reply:
Discussed in person: this is not a public api yet, and will have very limited usage here so we are going to leave it as is for now.